stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+		return prefs.units.weight == units::KG ? total_weight(d) / 1000.0
+						       : grams_to_lbs(total_weight(d));
+	}
+	std::vector<StatsOperation> supportedOperations() const override {
+		return { StatsOperation::Median, StatsOperation::Mean, StatsOperation::Sum };
+	}
+};
+
+// ============ Dive mode ============
+
+struct DiveModeBinner : public SimpleBinner<DiveModeBinner, IntBin> {
+	QString format(const StatsBin &bin) const override {
+		return QString(divemode_text_ui[derived_bin(bin).value]);
+	}
+	int to_bin_value(const dive *d) const {
+		int res = (int)d->dc.divemode;
+		return res >= 0 && res < NUM_DIVEMODE ? res : OC;
+	}
+};
+
+static DiveModeBinner dive_mode_binner;
+struct DiveModeType : public StatsVariableTemplate<StatsVariable::Type::Discrete> {
+	QString name() const override {
+		return StatsTranslations::tr("Dive mode");
+	}
+	QString diveCategories(const dive *d) const override {
+		int mode = (int)d->dc.divemode;
+		return mode >= 0 && mode < NUM_DIVEMODE ?
+			QString(divemode_text_ui[mode]) : QString();
+	}
+	std::vector<const StatsBinner *> binners() const override {
+		return { &dive_mode_binner };
+	}
+};
+
+// ============ Buddy (including dive guides) ============
+
+struct BuddyBinner : public StringBinner<BuddyBinner, StringBin> {
+	std::vector<QString> to_bin_values(const dive *d) const {
+		std::vector<QString> dive_people;
+		for (const QString &s: QString(d->buddy).split(",", SKIP_EMPTY))
+			dive_people.push_back(s.trimmed());
+		for (const QString &s: QString(d->divemaster).split(",", SKIP_EMPTY))
+			dive_people.push_back(s.trimmed());
+		return dive_people;
+	}
+};
+
+static BuddyBinner buddy_binner;
+struct BuddyType : public StatsVariableTemplate<StatsVariable::Type::Discrete> {
+	QString name() const override {
+		return StatsTranslations::tr("Buddies");
+	}
+	QString diveCategories(const dive *d) const override {
+		QString buddy = QString(d->buddy).trimmed();
+		QString divemaster = QString(d->divemaster).trimmed();
+		if (!buddy.isEmpty() && !divemaster.isEmpty())
+			buddy += ", ";
+		return buddy + divemaster;
+	}
+	std::vector<const StatsBinner *> binners() const override {
+		return { &buddy_binner };
+	}
+};
+
+// ============ Gas type, in 2%, 5%, 10% and 20% steps  ============
+// This is a bit convoluted: We differentiate between four types: air, pure oxygen, EAN and trimix
+// The latter two are binned in x% steps. The problem is that we can't use the "simple binner",
+// because a dive can have more than one cylinder. Moreover, the string-binner might not be optimal
+// because we don't want to format a string for every gas types, when there are thousands of dives!
+// Note: when the same dive has multiple cylinders that fall inside a bin, that cylinder will
+// only be counted once. Thus, depending on the bin size, the number of entries may change!
+// In addition to a binner with percent-steps also provide a general-type binner (air, nitrox, etc.)
+
```

Comment:
that gas binner is... crazy... when testing the code I kept trying to figure out how it bins my gases. Just wow.

Reply:
It is a bit brutal, indeed. There are two dimensions to this:
- You can bin coarsely by type and finely by percentage.
- The gas is selected depending on the variable (max o2, max he). Not that it is too useful, but maybe someone wants to plot bottom gas vs. deco gas?

Reply:
There are so many options there. I'm not convinced I fully understood everything you offer.